lib/expand/join.go

```diff
@@ -504,6 +507,29 @@ func (p *Peer) getAgent(opCtx operationContext) (*rpcserver.PeerServer, error) {
 
 func (p *Peer) run() error {
 	ctx, err := p.connect()
+
+	// schedule a cleanup function to fail the operation if this exits
+	// with error
+	defer func() {
+		if err == nil {
+			return
+		}
+		p.WithError(err).Warn("Peer is exiting with error.")
+		stopCtx, cancel := context.WithTimeout(context.Background(), defaults.AgentStopTimeout)
+		defer cancel()
+		if p.agent != nil {
+			p.Info("Stopping peer.")
+			if err := p.agent.Stop(stopCtx); err != nil {
+				p.WithError(err).Error("Failed to stop peer.")
+			}
+		}
+		if ctx != nil && ctx.Operation.ID != "" {
+			if err2 := ops.FailOperation(ctx.Operation.Key(), ctx.Operator, err.Error()); err2 != nil {
+				p.WithError(err2).Error("Failed to mark the operation as failed.")
```

Comment:
Add more fields to the log such as operation ID? Might be useful for log entries above too.